### DISCUSSION TITLE
Ragdoll Best Practices

### DISCUSSION THREAD

**Discussion by joshcoyne** [jrouwe/JoltPhysics#554] [Q&A] [answered]
Hello! I've integrated jolt physics into my engine and have started integrating the ragdoll support. I think I have it close, but I was wondering if there were any best practices I should consider? For example should the rigid body shapes overlap or not at all?

I've upload a picture of my rig in blender in base pose to help provide some visualization. I'm sure the rig isn't ideal (It's my first one!) but it seems to work mostly ok. 

Lastly, just wanna say this library is really awesome and it took me less than a day to have physical entities in my project bringing in the cmake project.

![Screenshot from 2023-05-31 19-17-27](https://github.com/jrouwe/JoltPhysics/assets/7210978/22c23d67-4b26-47f0-a015-59b7b1ea9275)

**Comment by jrouwe** [comment, marked as answer]
Hello,

Thanks!

Here are some tips:

* Make the rigid bodies of the limbs overlap so that nothing can get in between them. This can cause instabilities.
* Make sure you disable the collisions between the limbs and make sure that in the rest pose there are no internal collisions in the ragdoll. Internal collisions cause fighting between the constraint and the collision and not good for performance nor for stability (see RagdollSettings::DisableParentChildCollisions)
* Make sure that all your animations stay within the joint limits of the ragdoll. If you use e.g. motors to drive the ragdoll to a pose that's outside the limits you'll again have constraints fighting each other.
* For best stability: Masses of ragdoll bodies should decrease the further away you get from the root (see RagdollSettings::Stabilize)

**Reply by joshcoyne** [reply]
Thanks for the answer. That all makes sense and will be easy to incorporate on my side :)

**Reply by joshcoyne** [reply]
Hello! I am tuning more parameters and am a bit confused on how to setup the twist and plane axis for the swingtwistconstraint. Right now I have child bones in blender that I then derive a twist axis from, and then cross that with world z to get the plane axis. Does that sound appropriate to you? Maybe theres a simpler way to math it out but this is just for load :)

**Reply by jrouwe** [reply]
The normal and plane axis together form a basis that allows you to aim the constraint:

![SwingTwistConstraint](https://github.com/jrouwe/JoltPhysics/assets/1621693/4eff0f23-9aff-4ef9-933c-39704a877c77)

You can do twist cross world z to get a plane axis (note that you should take care that no joint is parallel to world z). However, unless min twist angle = max twist angle and normal cone angle = plane cone angle, the constraint limits are not symmetrical so the normal/plane axis should be user choices. A simple way would be to use the joint rotation of the default pose in blender: Artist tend to align 1 axis with the twist axis which leaves the other 2 axis to be used as normal/plane axis.

**Comment by joshcoyne** [comment]
awesome, that makes alot of sense, thank you :D
